teika/context.ml

```diff
@@ -1,225 +1,63 @@
-          let vars = Var.Map.remove var vars in
-          Ok (content, { loc; level; vars }))
-
-let enter_var_hole var =
-  Context
-    (fun data ->
-      let { loc; level; vars } = data in
-      match Var.Map.mem var vars with
-      | true ->
-          let desc = CError_duplicated_var_hole { var } in
-          Error (CError { loc; desc }, data)
-      | false ->
-          let info = V_hole { level } in
-          let vars = Var.Map.add var info vars in
-          Ok ((), { loc; level; vars }))
-
-let var_link var ~to_ =
-  (* TODO: path compression *)
-  Context
-    (fun data ->
-      let { loc; level; vars } = data in
-      (* TODO: we could track substitution locations here for LSP *)
-      match Var.Map.find_opt var vars with
-      | Some (V_hole { level = _ }) ->
-          let info = V_link { type_ = to_ } in
-          let vars = Var.Map.add var info vars in
-          Ok ((), { loc; level; vars })
-      | Some ((V_rigid _ | V_alias _ | V_link _) as info) ->
-          let desc = CError_invalid_var_link { var; info } in
-          Error (CError { loc; desc }, data)
-      | None ->
-          let desc = CError_unknown_var_link { var } in
-          Error (CError { loc; desc }, data))
-
-let var_lower var ~to_ =
-  Context
-    (fun data ->
-      let { loc; level; vars } = data in
-      (* TODO: we could track lowering locations here for LSP *)
-      match Var.Map.find_opt var vars with
-      | Some (V_hole { level = hole_level }) -> (
-          match Level.(hole_level < to_) with
-          | true ->
-              let desc = CError_lowering_to_higher_level { var } in
-              Error (CError { loc; desc }, data)
-          | false ->
-              let info = V_hole { level = to_ } in
-              let vars = Var.Map.add var info vars in
-              Ok ((), { loc; level; vars }))
-      | Some ((V_rigid _ | V_alias _ | V_link _) as info) ->
-          let desc = CError_invalid_var_lower { var; info } in
-          Error (CError { loc; desc }, data)
-      | None ->
-          let desc = CError_unknown_var_lower { var } in
-          Error (CError { loc; desc }, data))
-
-let var_alias var ~of_ =
-  Context
-    (fun data ->
-      let { loc; level; vars } = data in
-      (* TODO: we could track lowering locations here for LSP *)
-      match Var.Map.find_opt var vars with
-      | Some (V_hole { level = _ }) ->
-          (* TODO: check level is equal or higher than of_ *)
-          let info = V_alias { var = of_ } in
-          let vars = Var.Map.add var info vars in
-          Ok ((), { loc; level; vars })
-      | Some ((V_rigid _ | V_alias _ | V_link _) as info) ->
-          let desc = CError_invalid_var_alias { var; info } in
-          Error (CError { loc; desc }, data)
-      | None ->
-          let desc = CError_unknown_var_alias { var } in
-          Error (CError { loc; desc }, data))
+and error_desc = |
+
+module Subst_context = struct
+  type 'a subst_context = {
```

Comment:
This context is CPS, so that it avoid intermediary allocations and it can be easily optimized in the future. In fact if you try this code with flambda2 all this monadic stuff is removed.